### PR TITLE
Functional samples

### LINKED ISSUE #241: Feature Request: Allow sampling of log probs and logits for Likelihood.CLASSIFICATION
The code in `_glm_predictive_samples` always applies `torch.softmax` to the results under classification.

For numerical stability supporting `torch.log_softmax` here would be helpful. Similarly, it would be helpful if there was an easy way to obtain the logits without having to change self.likelihood intermittently.

Thanks,\
Andreas

### PR DESCRIPTION
**Pull request by wiseodd**
Closes #241.

**Use case:** computation of an arbitrary expectation w.r.t. $p(f(x) \mid x, \mathcal{D})$.

### CODE REVIEW

**Review by runame — APPROVED** [review]
This solution is nice because it is really flexible. I guess one small potential downside is that in this case you can't use `__call__` with another link function. Maybe this is even a good thing since using `functional_samples` is a lot more explicit than yet another argument to `__call__`.

File: laplace/baselaplace.py [hunk 1]
```diff
@@ -664,15 +664,15 @@ def _glm_forward_call(
                 "Prediction path invalid. Check the likelihood, pred_type, link_approx combination!"
             )
 
-    def _glm_predictive_samples(
+    def _glm_functional_samples(
```
[the remark this answers is not in the document]
Comment (wiseodd) [reply laplace/baselaplace.py:667]: I think for UX (or DX -- developer experience) purposes, it's better as an instance method. This way, it's coherent with `_nn_functional_samples` since it can't be made static.

PS. Seems like `@staticmethod` is not recommended everywhere I looked.[^1][^2]

[^1]: https://www.seporaitis.net/posts/2020/05/05/python-staticmethod-usage/
[^2]: https://google.github.io/styleguide/pyguide.html#2174-decision

[the remark this answers is not in the document]
Comment (runame) [reply laplace/baselaplace.py:667]: Fair points, thanks!

**Comment by wiseodd** [comment]
> I guess one small potential downside is that in this case you can't use __call__ with another link function. Maybe this is even a good thing since using functional_samples is a lot more explicit than yet another argument to __call__

I agree with your last statement.

I believe the original semantic for `__call__` is "making *predictions* based on the posterior obtained via `__fit__`". So, I'd say having an option in `__call__` to do an arbitrary expectation is ill-suited. Esp. so since we don't support an arbitrary loss function for computing the Hessian.

**Comment by wiseodd** [comment]
@aleximmer could you please double-check and merge?

**Review by aleximmer — APPROVED** [review]
Thanks for the improvements. LGTM.
